torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
   return output;
 }
 
+
+// =======================================TransformerImpl================================
+TransformerImpl::TransformerImpl(TransformerOptions options_ ) : options(std::move(options_)){
+  reset();
+}
+
+void TransformerImpl::reset() {
+  // set up encoder
+  if (options.custom_encoder().is_empty()) {
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerEncoder trans_encoder(TransformerEncoderOptions(
+      TransformerEncoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_encoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->encoder = AnyModule(trans_encoder);
+  }
+  else {
+    this->encoder = options.custom_encoder().clone();
```

Comment:
.clone() does a deep copy, right?

Reply:
A deep copy is what we want here. Better to make sure that it made a deep copy here.

Reply:
https://github.com/pytorch/pytorch/blob/367426494759ddde0896665ed55c6f9af2870cf0/torch/csrc/api/include/torch/nn/module.h#L98-L115

Confirmed that it is a deep copy